src/aleph/http/client.clj

```diff
@@ -143,21 +148,32 @@
         (cond
 
           (instance? HttpResponse msg)
-          (let [rsp msg]
-
-            (let [s (netty/buffered-source (netty/channel ctx) #(.readableBytes ^ByteBuf %) buffer-capacity)
+          (if (http/decoder-failed? msg)
+            (handle-decoder-failure ctx msg response-stream)
+            (let [rsp msg
+                  s (netty/buffered-source (netty/channel ctx) #(.readableBytes ^ByteBuf %) buffer-capacity)
                   c (d/deferred)]
+
               (reset! stream s)
               (reset! complete c)
               (s/on-closed s #(d/success! c true))
               (handle-response rsp c s)))
 
           (instance? HttpContent msg)
-          (let [content (.content ^HttpContent msg)]
-            (netty/put! (.channel ctx) @stream content)
-            (when (instance? LastHttpContent msg)
-              (d/success! @complete false)
-              (s/close! @stream)))
+          (if (http/decoder-failed? msg)
+            ;; note that we are most likely to get this when dealing
+            ;; with transfer encoding chunked
+            (if-let [s @stream]
+              (do
+                ;; flag that body was not completed succesfully
+                (d/success! @complete true)
+                (s/close! s))
+              (handle-decoder-failure ctx msg response-stream))
```

Comment:
It feels like we're duplicating a lot of logic here. Is there a way to simplify it? Maybe check for decoder failure first, and then have separate branches for the `Http*` types?

Reply:
I will try to improve it.

Reply:
https://github.com/clj-commons/aleph/pull/658/commits/6fc413ed4499b8c22aa29119503284410b142727